### PR TITLE
CA-399757: Add CAS style check for SR scan

### PR DESCRIPTION
**Pull request by Vincent-lau**
This is to reduce the chance of race between `SR.scan` and `VDI.db_introduce`. More details in the comment, but this won't fix the problem, but just makes it less likely to happen...

I have seen a few instances of this happening now in SXM tests, so need to prioritise fixing this

Trying to run some tests on this, but it's hard to know effective this is as storage migration tends to require lots of resources, and does not get run very often. Plus this issue is not very reproducible.

### CODE REVIEW

**Review by psafont — APPROVED** [review]
Can there be a situation where the two lists don't converge?

**Comment by Vincent-lau** [comment]
> Can there be a situation where the two lists don't converge?

Well then someone has to be constantly changing it under our feet, which I suppose wouldn't happen....

File: ocaml/xapi/xapi_sr.ml [hunk 1]
```diff
@@ -786,26 +786,40 @@ let scan ~__context ~sr =
...
+            in
+            let vs, sr_info =
+              C.SR.scan2 (Ref.string_of task)
+                (Storage_interface.Sr.of_string sr_uuid)
+            in
+            let db_vdis_after =
+              Db.VDI.get_records_where ~__context
+                ~expr:(Eq (Field "SR", Literal sr'))
+            in
+            if db_vdis_after <> db_vdis_before then
```
[the remark this answers is not in the document]
Comment (Vincent-lau) [reply ocaml/xapi/xapi_sr.ml:806]: The comparison isn't order invariant, which means it cares about the order? And if the order changes, we could see a different `db_vdis_after` (B) from `db_vdis_before` (A), and their difference is purely due to their order of elements. Where is the second A in the ABA problem?

**Comment by Vincent-lau** [comment]
> Could we mitigate the TOCTOU issue somewhat by holding the DB lock when computing `after`, such that we continue to holding it after we break the loop (fixpoint convergence), then do all the updates whilst holding the lock for the entirety (ensuring not to hold it during the scan), and then release it.
> 
> I don't know how expensive all the storage operations are in general or if this is the TOCTOU issue you mention, so this suggestion may be deeply destructive if you hold the lock for too long. This assumes the TOCTOU issue you're talking about is the potential to interleave changes between the breaking of the loop and the subsequent updates.

The TOCTOU problem happens when the db gets changed after we do the check `db_vdis_after <> db_vdis_before`, at which point we shouldn't really proceed.

So you are suggesting holding the lock before ` if db_vdis_after <> db_vdis_before then` and release it until after we finish updating the db? How do you hold a db lock for part of the operations `db_lock.lock db_lock.db_lock`?

File: ocaml/xapi/xapi_sr.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lindig) [reply ocaml/xapi/xapi_sr.ml:806]: I agree that we should exclude non termination by having an upper limit on iterations and maybe an internal error.

[the remark this answers is not in the document]
Comment (Vincent-lau) [reply ocaml/xapi/xapi_sr.ml:806]: Agreed, what I am worried about is callers of SR.scan might not be prepared to handle this properly as previously they wouldn't expect SR.scan to throw an exception. Then perhpas we need to change the semantics of the callers (and how they deal with the exceptions).

**Comment by robhoes** [comment]
> Yes I am worried about deadlocking if we hold this lock. It could be that SR.scan is run on a pool member, and it will hold the lock for a period, while invoking updates such as Db.SR.... which will be run on the coordinator, and because the member is holding the lock, this will cause coordinator never able to acquire the db lock at all. And this is just one case, there might be other cases of dead locking as well.

This function can run on any host (e.g. a pool member for a local SR), but the coordinator runs the database and only the coordinator has access to DB locks.

File: ocaml/xapi/xapi_sr.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (robhoes) [reply ocaml/xapi/xapi_sr.ml:806]: I also think we should add a max number of iterations and raise an internal error if exceeding it. Callers will not expect it, but well, it's an internal error that we don't expect to see in practice, assuming the number of iterations is large enough.

**Comment by robhoes** [comment]
I think this is fine after adding the iteration limit, as discussed above.

**Review by robhoes — APPROVED** [review]

**Pull request merged**